File: src/main/java/com/jcloisterzone/reducers/DeployMeeple.java
package com.jcloisterzone.reducers;

import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.event.MeepleDeployed;
import com.jcloisterzone.event.PlayEvent.PlayEventMeta;
import com.jcloisterzone.feature.Structure;
import com.jcloisterzone.figure.DeploymentCheckResult;
import com.jcloisterzone.figure.Meeple;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.capability.SoloveiCapability;
import com.jcloisterzone.game.capability.VodyanoyCapability;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import io.vavr.Tuple2;
import io.vavr.collection.LinkedHashMap;
import io.vavr.collection.Set;

public class DeployMeeple implements Reducer {

    private final Meeple meeple;
    private final FeaturePointer fp;

    public DeployMeeple(Meeple meeple, FeaturePointer fp) {
        this.meeple = meeple;
        this.fp = fp;
    }

    @Override
    public GameState apply(GameState state) {
        Structure feature = state.getStructure(fp);
        if (feature == null) {
            throw new IllegalArgumentException("There is no feature on " + fp);
        }

        DeploymentCheckResult check = meeple.isDeploymentAllowed(state, fp, feature);
        if (!check.result) {
            throw new IllegalArgumentException(check.error);
        }

        LinkedHashMap<Meeple, FeaturePointer> deployedMeeples = state.getDeployedMeeples();
        FeaturePointer movedFrom = deployedMeeples.get(meeple).getOrNull();
        state = state.setDeployedMeeples(deployedMeeples.put(meeple, fp));
        state = state.appendEvent(
            new MeepleDeployed(PlayEventMeta.createWithActivePlayer(state), meeple, fp, movedFrom)
        );

        if (state.getCapabilities().contains(SoloveiCapability.class)) {
        	SoloveiCapability soloveiCapability = state.getCapabilities().get(SoloveiCapability.class);
    	    for (Tuple2<Position, PlacedTile> t : soloveiCapability.getSoloveisPlacedTiles(state)) {
                state = soloveiCapability.onTilePlaced(state, t._2);
            };
        }
        if (state.getCapabilities().contains(VodyanoyCapability.class)) {
        	VodyanoyCapability vodyanoyCapability = state.getCapabilities().get(VodyanoyCapability.class);
    	    for (Tuple2<Position, PlacedTile> t : vodyanoyCapability.getVodyanoysPlacedTiles(state)) {
                state = vodyanoyCapability.onTilePlaced(state, t._2);
            };
        }
        return state;
    }

}

File: src/main/ts/com/jcloisterzone/reducers/DeployMeeple.ts
import type { FeaturePointer } from "../board/pointer/FeaturePointer.js";
import { PlayEventMeta } from "../event/PlayEvent.js";
import { MeepleDeployed } from "../event/MeepleDeployed.js";
import { Follower } from "../figure/Follower.js";
import type { Meeple } from "../figure/Meeple.js";
import { KeepBuildingCapability } from "../game/capability/KeepBuildingCapability.js";
import { Flag } from "../game/state/Flag.js";
import type { GameState } from "../game/state/GameState.js";
import { AbstractUndeploy } from "./AbstractUndeploy.js";

/** Deploys a meeple onto a feature pointer. */
export class DeployMeeple extends AbstractUndeploy {
  constructor(
    private readonly meeple: Meeple,
    private readonly fp: FeaturePointer,
  ) {
    super();
  }

  apply(state: GameState): GameState {
    const feature = state.getStructure(this.fp);
    if (feature === null) {
      throw new Error("There is no feature on " + this.fp);
    }

    const check = this.meeple.isDeploymentAllowed(state, this.fp, feature);
    if (!check.result) {
      throw new Error(check.error ?? "Deployment not allowed");
    }

    // Keep Building (coop variant), condition (b): a follower (phantom/wagon included;
    // the barn is a Special, hence excluded) occupies a previously unoccupied feature.
    // Checked before the deploy is recorded, so `isOccupied` reflects the prior state.
    // Once earned, the flag is never revoked this turn (scoring return / dragon don't undo it).
    if (
      this.meeple instanceof Follower &&
      !state.hasFlag(Flag.COOP_CONDITION_MET) &&
      state.hasCapability(KeepBuildingCapability) &&
      !feature.isOccupied(state)
    ) {
      state = state.addFlag(Flag.COOP_CONDITION_MET);
    }

    const deployedMeeples = state.getDeployedMeeples();
    const movedFrom = deployedMeeples.get(this.meeple).getOrNull();
    state = state.setDeployedMeeples(deployedMeeples.put(this.meeple, this.fp) as typeof deployedMeeples);
    state = state.appendEvent(
      new MeepleDeployed(PlayEventMeta.createWithActivePlayer(state), this.meeple, this.fp, movedFrom),
    );

    if (movedFrom !== null && this.meeple instanceof Follower) {
      state = this.undeployLonelySpecials(state, this.meeple, movedFrom, true);
    }
    return state;
  }
}
